File: code/py/ch08/08.02.py
## 8.2 重回帰分析

import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score, LeaveOneOut

my_url = ('https://raw.githubusercontent.com/taroyabuki'
          '/fromzero/master/data/wine.csv')
my_data = pd.read_csv(my_url)
X, y = my_data.drop(columns=['LPRICE2']), my_data['LPRICE2']

my_model = LinearRegression().fit(X, y)

my_model.intercept_
#> -12.145333576510417

pd.Series(my_model.coef_,
          index=X.columns)
#> WRAIN      0.001167
#> DEGREES    0.616392
#> HRAIN     -0.003861
#> TIME_SV    0.023847
#> dtype: float64

my_test = [[500, 17, 120, 2]]
my_model.predict(my_test)
#> array([-1.49884253])

y_ = my_model.predict(X)

mean_squared_error(y_, y)**0.5
#> 0.2586166620130621 # RMSE（訓練）

my_model.score(X, y)
#> 0.8275277990052154 # 決定係数1

np.corrcoef(y, y_)[0, 1]**2
#> 0.8275277990052158 # 決定係数6

my_scores = cross_val_score(my_model, X, y,
                            cv=LeaveOneOut(),
                            scoring='neg_mean_squared_error')
(-my_scores.mean())**0.5
#> 0.32300426518411957 # RMSE（検証）

### 8.2.1 補足：行列計算による再現

import numpy as np
M = np.matrix(X.assign(b0=1))
b = np.linalg.pinv(M) @ y
pd.Series(b,
    index=list(X.columns) + ['b0'])
#> WRAIN       0.001167
#> DEGREES     0.616392
#> HRAIN      -0.003861
#> TIME_SV     0.023847
#> b0        -12.145334
#> dtype: float64

File: code/R/ch08/08.02.R
## 8.2 重回帰分析

library(caret)
library(tidyverse)
my_url <- str_c("https://raw.githubusercontent.com/taroyabuki",
                "/fromzero/master/data/wine.csv")
my_data <- read_csv(my_url)

my_model <- train(form = LPRICE2 ~ WRAIN + DEGREES + HRAIN + TIME_SV,
                  data = my_data,
                  method = "lm",
                  trControl = trainControl(method = "LOOCV"))

coef(my_model$finalModel) %>%
  as.data.frame
#>                         .
#> (Intercept) -12.145333577
#> WRAIN         0.001166782
#> DEGREES       0.616392441
#> HRAIN        -0.003860554
#> TIME_SV       0.023847413

my_test <- data.frame(
  WRAIN = 500, DEGREES = 17,
  HRAIN = 120, TIME_SV = 2)
my_model %>% predict(my_test)
#>         1
#> -1.498843

y  <- my_data$LPRICE2
y_ <- my_model %>% predict(my_data)

RMSE(y_, y)
#> [1] 0.2586167 # RMSE（訓練）

R2(pred = y_, obs = y,
   form = "traditional")
#> [1] 0.8275278 # 決定係数1（訓練）

R2(pred = y_, obs = y,
   form = "corr")
#> [1] 0.8275278 # 決定係数6（訓練）

my_model$results
#>   intercept      RMSE  Rsquared       MAE
#> 1      TRUE 0.3230043 0.7361273 0.2767282

### 8.2.1 補足：行列計算による再現

M <- my_data[, -1] %>%
  mutate(b0 = 1) %>% as.matrix
b <- MASS::ginv(M) %*% y
matrix(b,
       dimnames = list(colnames(M)))
#>                  [,1]
#> WRAIN     0.001166782
#> DEGREES   0.616392441
#> HRAIN    -0.003860554
#> TIME_SV   0.023847413
#> b0      -12.145333577
